src/pages/data-first-and-data-last-a-comparison.md

```diff
@@ -0,0 +1,474 @@
+```
+
+Note the difference: in this case, the compiler assumes the type of `words`, `list(string)`, is the truth, and then it fails when the callback returns an `int` type. Note also how the error message is simpler: **the compiler is not matching `list(int)` against `list(string)`** like in the first case, **but `int` against `string`**. Because it is operating already "inside" the callback, it can match the type of `n` (`string`) against the type of `1` (`int`).
+
+This might not seem a big deal in this small example, but for real applications where the functions and data become more complex, the errors can become quite more cryptic with the data-last approach, because the compiler is assuming the source of truth is that of the "lifted" types of the callback: list, maps, options and any other "wrapping" types that are used in those callbacks.
+
+### Swimming upstream
+
+As we are seeing, data-first can be more ergonomic towards inference, considering the type checker processes code from left to right. In a few situations, the compiler might not be able to infer the types of data-last functions properly.
+
+Let's say we have a module `User` with the following implementation:
+
+```reason
+module User = {
+  type t = {
+    name: string,
+    age: int,
+  };
+  let admins = [
+    {name: "Jane", age: 30},
+    {name: "Manuel", age: 72},
+    {name: "Lin", age: 54},
+  ];
+}
+```
+
+Now, outside this module, we want to get a list with the ages of the `admins` users. We use OCaml standard library `List` function `List.map`:
+
+```reason
+module User = { ... }
+let ages = List.map(u => u.age, User.admins);
+                           ^^^
+```
+
+```
+The record field age can't be found.
+If it's defined in another module or file, bring it into scope by:
+  - Annotating it with said module name: let baby = {MyModule.age: 3}
+  - Or specifying its type: let baby: MyModule.person = {age: 3}
+```
+
+> [Try it yourself](https://reasonml.github.io/en/try?rrjsx=true&ocaml=LYewJgrgNgpgBAVQM4wE5wLwCg5yQF1QgGN8dc58BPAB3n0zgG9yK4A7AQ2BgC49CAS3YBzANysKnEXzjD8YgL6TYDTmGDCkmSbgDaTDt3gY4ACjMAiAFKd2MSwEo9AAVQxOSEOwB0qTgDuAPpQgvhonFBwNnYOALqOYnDSJnAAzAAMcIoSbLiGXDyMFpYAsnYQMFBOru6e3n6BIWERUWUVVZYJSSmMAOwATNm5eQXGxVYAMsI1bh5evv7BoeH+bdPsXYnJMowArAAs2XHkMOxgWKo7MNqm0wQ+wJw05gBmEOxwELgAtAB8Xx8KUciBQqCBGi0QA).
+
+Whoa whoa there compiler... "_annotating it_"? "_specifying its type_"?! I was promised OCaml had such a powerful inference engine that I would never need to write any more type annotations! 😄
+
+Jokes aside, it seems the compiler can't figure out that we want to get the value of the field `age` from the record of type `User.t`, even if it has `User.admins` right next to the callback, and it knows it has type `list(User.t)`.
+
+We can solve the problem by adding a type annotation, as suggested by the compiler error:
+
+```reason
+let ages = List.map((u: User.t) => u.age, User.admins);
```

Comment:
I recommend prefixing the field with the module name, as per the first suggestion: `let ages = List.map(u => u.User.age, User.admins);` It's more 'lightweight' than a full-fledged type annotation and uniformly works in every equivalent situation, like record literals, and variant data constructors.